### DISCUSSION TITLE
Run 2 programs in succession

### DISCUSSION THREAD

**Discussion by PatLub85** [micropython/micropython#12672] [General]
I would like to know if it is possible to make an ESP works in the following way:
- boot, launch of program 1 (main.py?)
- presses a push button, launches program 2, with release of program 1 and dependencies
- presses a push button, launches program 1, with release of program 2 and dependencies
- etc ...
The idea is that each program has the maximum  memory.
But how to keep the previous state of the push button?
Thanks in advance

**Comment by 2dof** [comment]
Read/do some examples from [micropython-async](https://github.com/peterhinch/micropython-async/blob/master/v3/docs/TUTORIAL.md)  about  asyncio tasks.
best start form  libs for [push/switch button](https://github.com/peterhinch/micropython-async/blob/master/v3/docs/DRIVERS.md)

your program  button switching function is just description of J-K flip-flop  functionality (without clock) . in simplest  pure code realisation will be like (as button calback)
(pseudo code)
```
state_flag =  False   # some global flaf

def  pushBtn_callbac(): 
      global state_flag
       
      state_flag = not state_flag

     if state_flag:
       # terminate program 1 (task cancel) and safe program states (if needed) 
       # run program 2  (  see  asyncio.create_task)
     else:
       # terminate program 2 
       # run program 1 
```

 for memory cleaning use [garbage collector](https://docs.micropython.org/en/latest/library/gc.html)

**Reply by glenn20** [reply]
One option is to reboot the device completely with a different app to run each time. You need a way to persist the choice of app to run between reboots. You could do that with the filesystem (need to think about wear on the flash memory), [esp32.NVS storage](https://docs.micropython.org/en/v1.15/library/esp32.html#non-volatile-storage) (which mitigates impact of wear) or `machine.RTC.memory`. If you have any other state to persist across reboots, you will need to handle that too.

**Reply by PatLub85** [reply]
This seems like a good idea to me. I will test machine.RTC.memory.
Thank you for your help

**Reply by glenn20** [reply]
NP. Oh, one thing - `machine.RTC.memory` is currently undocumented. You can use it to write a `bytes` or `bytearray` object into RTC memory and read it back after a reboot (of course this is lost on a power cycle). An example of usage (untested code):

```py
import machine

rtc = machine.RTC()
app_names = ("app_1", "app_2", "app_3")
app_num = (rtc.memory() or b"\x00")[0]  # Select first app if no value stored
next_app_num = (app_num + 1) % len(app_names)
rtc.memory(bytearray([next_app_num]))  # Save the next app num into RTC memory

# Load interrupt handlers for pins with handlers which call machine.soft_reset() on 
# button press or flag the app to cleanup and then reset

__import__(app_names[app_num])  # Import the app
```

Note that RTC memory is cleared on `machine.hard_reset()`, but you can use `machine.soft_reset()` or `machine.deepsleep(1)`.

If you want to save some other state between resets you could encode that into the bytearray with [`json`](https://docs.micropython.org/en/latest/library/json.html), [`struct`](https://docs.micropython.org/en/latest/library/struct.html#module-struct) or [`uctypes`](https://docs.micropython.org/en/latest/library/uctypes.html). The max length of data you can store is 2048 bytes (but can be changed by recompiling micropython).

**Reply by PatLub85** [reply]
Thanks Glenn,
I think you have the solution to my problem. I will test this next week

**Reply by PatLub85** [reply]
I don't understand why the '% len(app_names)'

**Reply by PatLub85** [reply]
sorry, I understood. This allows next_app_num to return to 0
